src/librustc_mir/transform/check_unsafety.rs

```diff
@@ -679,7 +678,7 @@ pub fn check_unsafety<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId) {
     }
 
     let mut unsafe_blocks: Vec<_> = unsafe_blocks.into_iter().collect();
-    unsafe_blocks.sort();
+    unsafe_blocks.sort_by_cached_key(|(hir_id, _)| tcx.hir().hir_to_node_id(*hir_id));
```

Comment:
I'm not sure if this is the best way to tackle it, but the regular sort now breaks the order of error messages in 2 "unnecessary unsafe" test cases.